provider/coredns/coredns.go

```diff
@@ -148,10 +177,53 @@ func (c etcdClient) DeleteService(ctx context.Context, key string) error {
 	ctx, cancel := context.WithTimeout(ctx, etcdTimeout)
 	defer cancel()
 
+	if owned, err := c.isOwnedBy(ctx, key); err != nil {
+		return err
+	} else if !owned {
+		return fmt.Errorf("key %q is not owned by this service", key)
+	}
+
 	_, err := c.client.Delete(ctx, key, etcdcv3.WithPrefix())
 	return err
 }
 
+func (c etcdClient) isOwnedBy(ctx context.Context, key string) (bool, error) {
+	if !c.strictlyOwned {
+		return true, nil
+	}
+
+	r, err := c.client.Get(ctx, key)
+	switch {
```

Comment:
This does not look like go code. Let's not have switch, especially when there is error handling

```
if err != nil {
		return false, fmt.Errorf("etcd get %q: %w", key, err)
	}
	if r == nil || len(r.Kvs) == 0 {
		// Key missing -> treat as owned (safe to create)
		return true, nil
	}
```